### PR TITLE
[AIRFLOW-3123] Allow nested use of  DAG as a context manager

### PR DESCRIPTION
**Pull request by newtonle**
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-3123
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:
Implemented a stack to manage the DAG context. Entering a DAG context pushes the past context on the stack. Exiting pops the stack to reset the context. This allows arbitrary nesting of contexts, particularly if the same DAG object is used in the nesting.

### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:
Modified `test_dag_as_context_manager` in `models.py` to capture this condition.

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [ ] In case of new functionality, my PR adds documentation that describes how to use it.
  - When adding new operators/hooks/sensors, the autoclass documentation generation needs to be added.

### Code Quality

- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

### CODE REVIEW

**Comment by newtonle** [comment]
@aoen

File: airflow/models.py [hunk 1]
```diff
@@ -3389,7 +3389,7 @@ def __init__(
         self.on_success_callback = on_success_callback
         self.on_failure_callback = on_failure_callback
 
-        self._context_manager_set = False
+        self._old_context_manager_dags = []
```
Comment (aoen) [inline airflow/models.py:3394]: Nit: How about just context_manager_dags since it includes the current one?

Comment (newtonle) [reply airflow/models.py:3394]: Actually this doesn't contain the current one. It just contains what was in `_CONTEXT_MANAGER_DAG` before it is set to the current DAG. The `old` naming convention was existing, and the only difference here is making into a list.

**Review by ashb — APPROVED** [review]
I'm not sure why you'd want to do this in practice, but code looks sensible :)

**Comment by bcb** [comment]
@newtonle curious to know why you'd need to nest this context manager?

**Comment by newtonle** [comment]
> @newtonle curious to know why you'd need to nest this context manager?

We've built some constructs that can build DAGs, often using the DAG context manager. The problem is that user can use these constructs within their own DAG definition files, often inside another context manager. The problem with having only the last context manager saved is that entering a second one will lose history of the first. The stack-based approach makes this more robust with minimal additional overhead.

**Pull request merged**